bvm/ballerina-core/src/main/java/org/ballerinalang/model/util/serializer/JsonDeserializer.java

```diff
@@ -0,0 +1,442 @@
+            }
+        }
+
+        // try reflective reconstruction
+        if (object == null) {
+            Object emptyInstance = createInstance(jBMap, targetType);
+            // add the obj reference here so that inner deserialization may refer this object
+            addObjReference(jBMap, emptyInstance);
+            object = deserializeInto(emptyInstance, jBMap, targetType);
+
+            // check to make sure deserializeInto returns the populated 'emptyInstance'.
+            // It's important  that it does not create own objects as it may interfere with
+            // handling of existing references.
+            if (object != emptyInstance) {
+                throw new BallerinaException("Internal error: deserializeInto should not create own objects.");
+            }
+        }
+
+        // execute java.io.Serializable.readResolve if available
+        Object resolved = readResolve(jBMap, targetType, object);
+        if (resolved != null) {
+            return resolved;
+        }
+        return object;
+    }
+
+    @SuppressWarnings("unchecked")
+    private Object deserializeInto(Object instance, BMap<String, BValue> jsonNode, Class<?> targetType) {
+        BValue payload = jsonNode.get(JsonSerializerConst.PAYLOAD_TAG);
+        BValue type = jsonNode.get(JsonSerializerConst.TYPE_TAG);
+        if (type != null) {
+            String objType = type.stringValue();
+            switch (objType) {
+                case JsonSerializerConst.MAP_TAG:
+                    return deserializeMap((BMap<String, BValue>) payload, (Map) instance);
+                case JsonSerializerConst.LIST_TAG:
+                    return deserializeList((BRefValueArray) payload, targetType, (List) instance);
+                case JsonSerializerConst.ENUM_TAG:
+                    return instance;
+                case JsonSerializerConst.ARRAY_TAG:
+                    return deserializeBRefValueArray((BRefValueArray) payload, instance);
+                default:
+                    // no op
+            }
+        }
+
+        if (payload == null) {
+            payload = jsonNode;
+        }
+        return deserializeReflectively(instance, (BMap<String, BValue>) payload);
+    }
+
+    private String resolveTargetTypeName(Class<?> targetType, BMap<String, BValue> jBMap) {
+        BValue jType = jBMap.get(JsonSerializerConst.TYPE_TAG);
+        if (jType != null) {
+            return jType.stringValue();
+        } else if (targetType != null) {
+            return targetType.getName();
+        }
+        return null;
+    }
+
+    public void addObjReference(BMap<String, BValue> jBMap, Object object) {
+        BInteger objId = (BInteger) jBMap.get(JsonSerializerConst.ID_TAG);
+        if (objId != null) {
+            identityMap.put(objId.intValue(), object);
+        }
+    }
+
+    private Object findExistingReference(BMap<String, BValue> jBMap) {
+        if (jBMap.hasKey(JsonSerializerConst.EXISTING_TAG)) {
```

Comment:
use the early exit pattern. i.e:
```java
if (!jBMap.hasKey()) {
    return null;
}
```

Then we can avoid unnecessary indentations.